### PR TITLE
ci: implement flux install/upgrade regression test pipeline

### PR DESCRIPTION
**Pull request by yangchiu**
ci: implement flux install/upgrade regression test pipeline

For https://github.com/longhorn/longhorn/issues/3987 https://github.com/longhorn/longhorn/issues/6343 https://github.com/longhorn/longhorn/issues/5936

Signed-off-by: Yang Chiu <yang.chiu@suse.com>

### CODE REVIEW

**Comment by yangchiu** [comment]
> This pull request is now in conflicts. Could you fix it @yangchiu? 🙏

conflicts resolved and ready for review. cc @longhorn/qa

File: manager/integration/deploy/test.yaml [hunk 1]
```diff
@@ -40,6 +40,12 @@ rules:
 - apiGroups: ["longhorn.io"]
   resources: ["supportbundles", "supportbundles/status"]
   verbs: ["*"]
+- apiGroups: ["source.toolkit.fluxcd.io"]
+  resources: ["helmrepositories"]
+  verbs: ["*"]
+- apiGroups: ["helm.toolkit.fluxcd.io"]
+  resources: ["helmreleases"]
+  verbs: ["*"]
```
Comment (innobead) [inline manager/integration/deploy/test.yaml:48]: We can just use cluster admin role, so we don't need to worry about the permission later for any kind of tests.

This is another topic, so can be tackled in another ticket.

File: manager/integration/tests/test_upgrade.py [hunk 2]
```diff
@@ -243,6 +270,8 @@ def test_upgrade(longhorn_upgrade_type,
     rancher_secret_key = rancher_secret_key
     rancher_chart_install_version = rancher_chart_install_version
     longhorn_repo = longhorn_repo
+    flux_helm_chart_url = flux_helm_chart_url
+    flux_helm_chart_version = flux_helm_chart_version
```
Comment (innobead) [inline manager/integration/tests/test_upgrade.py:274]: Why we need this kind of self assignment? it seems no any effect?

File: manager/integration/Dockerfile [hunk 3]
```diff
@@ -21,7 +21,10 @@ RUN curl -sO https://storage.googleapis.com/kubernetes-release/release/$KUBECTL_
     wget -q https://releases.hashicorp.com/terraform/${TERRAFORM_VERSION}/terraform_${TERRAFORM_VERSION}_linux_${ARCH}.zip && \
     unzip terraform_${TERRAFORM_VERSION}_linux_${ARCH}.zip && rm terraform_${TERRAFORM_VERSION}_linux_${ARCH}.zip && \
     mv terraform /usr/bin/terraform && \
-    chmod +x /usr/bin/terraform
+    chmod +x /usr/bin/terraform && \
+    curl -s -o flux_install.sh https://fluxcd.io/install.sh && \
```
Comment (innobead) [inline manager/integration/Dockerfile:25]: I assume this is always the latest and stable version of flux?

File: pipelines/flux/scripts/longhorn-setup.sh [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+      UPGRADE_LH_ENGINE_IMAGE="longhornio/longhorn-engine:${LONGHORN_TRANSIENT_VERSION}"
+      run_longhorn_upgrade_test
+      LONGHORN_UPGRADE_TYPE="from_transient"
+      LONGHORN_UPGRADE_TEST_POD_NAME="longhorn-test-upgrade-from-transient"
+    fi
+    FLUX_HELM_CHART_URL="${HELM_CHART_URL}"
+    FLUX_HELM_CHART_VERSION="${LONGHORN_INSTALL_VERSION}"
+    UPGRADE_LH_ENGINE_IMAGE="longhornio/longhorn-engine:${LONGHORN_INSTALL_VERSION}"
+    run_longhorn_upgrade_test
+    run_longhorn_test
```
Comment (innobead) [inline pipelines/flux/scripts/longhorn-setup.sh:60]: Does this mean we don't use flux to do the upgrade testing but rather relying on `run_longhorn_upgrade_test`?

**Review by innobead — CHANGES_REQUESTED** [review]
Lood good for the flux part, but some questions need to resolve/answer first.

File: manager/integration/deploy/test.yaml (hunk 1 above)
Comment (yangchiu) [reply manager/integration/deploy/test.yaml]: Refined to use `ClusterRole cluster-admin`.

File: manager/integration/tests/test_upgrade.py (hunk 2 above)
Comment (yangchiu) [reply manager/integration/tests/test_upgrade.py]: Removed.

File: manager/integration/Dockerfile (hunk 3 above)
Comment (yangchiu) [reply manager/integration/Dockerfile:25]: Yes, it always gets the latest release version.

File: pipelines/flux/scripts/longhorn-setup.sh (hunk 4 above)
Comment (yangchiu) [reply pipelines/flux/scripts/longhorn-setup.sh:60]: `run_longhorn_upgrade_test` is just a function, it applies the `test.yaml` with desired configuration and starts the test pod, and the test pod is who performing the upgrade test in `pytest`, and `pytest`'s upgrade test case will determine which method should be used to perform the upgrade according to the configuration. In this case, it's the configuration for `flux`, so `pytest`'s upgrade test case will call the `flux` related shell script to perform some `flux` operations.

Other upgrade methods do the same thing. The only difference is the configuration.

**Review by innobead — APPROVED** [review]
LGTM

**Pull request merged**